service/database_service_configuration.go

```diff
@@ -1,55 +1,251 @@
+// are cloud managed MySQL or PostgreSQL databases.
+//
+// Supported database protocols are: `mysql` and `postgres`. For upstream authentication, supported auth types
+// are: `username_password` and `iam`. When using IAM authentication, the client must provide AWS credentials,
+// AWS region and a username. You can provide an optional CA certificate to verify the RDS database server's
+// certificate.
+type AwsRdsDatabaseServiceConfiguration struct {
+	HostnameAndPort
+
 	DatabaseProtocol   string `json:"protocol"`
 	AuthenticationType string `json:"authentication_type"`
+
+	UsernameAndPassword *AwsRdsUsernameAndPasswordAuthConfiguration `json:"username_and_password_auth_configuration,omitempty"`
+	Iam                 *AwsRdsIamAuthConfiguration                 `json:"iam_auth_configuration,omitempty"`
 }
 
-// CloudSqlDatabaseServiceConfiguration
-type CloudSqlDatabaseServiceConfiguration struct {
-	UsingConnector bool `json:"using_cloudsql_connector"`
+// GoogleCloudSqlDatabaseServiceConfiguration represents service configuration for Google Cloud SQL databases.
+// Google Cloud SQL databases are cloud managed MySQL or PostgreSQL databases.
+//
+// Border0 currently supports two ways of connecting to Google Cloud SQL databases: with and without the Cloud SQL Connector.
+// Use the corresponding configuration fields to configure the upstream connection.
+type GoogleCloudSqlDatabaseServiceConfiguration struct {
+	CloudSqlConnectorEnabled bool `json:"cloudsql_connector_enabled"`
+
+	Connectorless *GoogleCloudSqlConnectorlessConfiguration `json:"connectorless_configuration,omitempty"`
+	Connector     *GoogleCloudSqlConnectorConfiguration     `json:"connector_configuration,omitempty"`
+}
+
+// =======================================================================================
+// Configurations specifically made for Google Cloud SQL
+// - connectorless: without cloud sql connector
+// - connector: with cloud sql connector
+// =======================================================================================
+
+// GoogleCloudSqlConnectorlessConfiguration represents service configuration for Google Cloud SQL databases that will
+// be connected to the upstream _WITHOUT_ using the Cloud SQL Connector.
+//
+// Supported database protocol is: `mysql`. For upstream authentication, supported auth types are: `username_password`,
+// and `tls`. When using TLS authentication, the client must provide a username, a password, a client certificate and a
+// client key.
+type GoogleCloudSqlConnectorlessConfiguration struct {
+	HostnameAndPort
+
+	DatabaseProtocol   string `json:"protocol"`
+	AuthenticationType string `json:"authentication_type"`
+
+	UsernameAndPassword *DatabaseUsernameAndPasswordAuthConfiguration `json:"username_and_password_auth_configuration,omitempty"`
+	Tls                 *DatabaseTlsAuthConfiguration                 `json:"tls_auth_configuration,omitempty"`
```

Comment:
tlsAuthConfiguration might be more explicit? I guess struct name does not matter.

Reply:
Yeah. I try to avoid repetitive naming. In this case, type is also part of the name, and in the type, it already includes `AuthConfiguration` suffix.

For coding style, I generally follow https://google.github.io/styleguide/go and https://github.com/golang/go/wiki/CodeReviewComments. In https://google.github.io/styleguide/go/decisions#repetition

> Names that include information from their surrounding context often create extra noise without benefit. The package name, method name, type name, function name, import path, and even filename can all provide context that automatically qualifies all names within.

Reply:
That's fine, i agree, but we should at least have some suffix on these field names to imply that they are mutually exclusive fields. 

E.g. having `AuthConfiguration` suffix helps with that, but there are other suffixes that may be more appropriate as per the style guide you posted.

What I'm saying is 

`UsernameAndPasswordAuthConfig` and `TlsAuthConfig` are clearly mutually exclusive (from the name) because they have the same prefix.

Maybe just `TlsAuth` and `UsernameAndPasswordAuth`? I still prefer the longer names for explicitness.

Reply:
Agreed on `TlsAuth` and `UsernameAndPasswordAuth`. It makes it clearer. I will make the change.